File: src/orchestradirector.rs
use crate::utils::respond_to_uci;
#[macro_export]
macro_rules! debug {
    () => {
        false
    };
}

use crate::timer::Timer;
use crate::utils::split_fen_moves;
use crate::board::Board;
use crate::utils::COLOR::{BLACK, WHITE};
use crate::engine::{new_engine, Engine};

pub struct OrchestraDirector {
    pub eng: Engine,
    timer: Timer,

}

pub fn new_orchestra_director() -> OrchestraDirector {
    OrchestraDirector {
        eng: new_engine(Board::empty_board()),
        timer: Timer::new_timer(),

    }
}

impl OrchestraDirector {
    pub(crate) fn init_startpos(&mut self) {
        self.eng.board.from_startpos();
    }

    pub fn handle_command(&mut self, command: &str, options: &str) {
        match command {
            "uci" => self.uci_handle_uci(),
            "isready" => self.uci_handle_isready(),
            "ucinewgame" => {} // Do nothing for "ucinewgame"
            "position" => self.uci_handle_position(options),
            "go" => self.uci_handle_go(options),
            "stop" => self.uci_handle_stop(),
            "quit" => self.uci_handle_quit(),
            "setoption" => {} // Do nothing for "setoption"
            _ => {
                if debug!() {
                    panic!("NotImplementedError: {} {}", command, options);
                }
            }
        }
    }

    fn uci_handle_uci(&self) {
        respond_to_uci("id name Orchestra");
        respond_to_uci("id author Dario & Mattia");
        respond_to_uci("uciok");
    }

    fn uci_handle_position(&mut self, options: &str) {
        // todo: we re-initialize everything here, so we have a 400ms overhead for magics + nnue after the position command
        // find a way to re-use the nnue (especially) and magics
        if options.starts_with("startpos") { // todo: review this because the string editing is done in two different places
            self.init_startpos();
            let w = options.split("moves").collect::<Vec<_>>();
            if w.len() > 1 {
                self.execute_moves(w[1]);
                self.eng.moves_loaded = w[1].to_string();
            }
            self.eng.position_loaded = "startpos".to_string();
        } else {
            let (fen, moves) = split_fen_moves(options);
            self.init_from_fen(&fen);
            self.execute_moves(&moves);
        }
    }

    fn execute_moves(&mut self, param: &str) {
        let moves: Vec<&str> = param.split_whitespace().collect();
        for mov_str in moves {
            let mov = self.eng.board.move_from_str(mov_str);
            self.eng.board.make_move(mov);
        }
    }

    fn init_from_fen(&mut self, fen: &str) {
        // todo: review this because the string editing is done in two different places
        // fen = options[options.find("[") + 1:options.find("]")]
        self.eng.board.from_fen(fen);
    }

    fn uci_handle_isready(&self) {
        respond_to_uci("readyok");
    }

    fn uci_handle_go(&mut self, options: &str) {
        println!("{}", options);

        let op_list: Vec<&str> = options.split_whitespace().collect();
        let mut i = 0;
        let mut movetime: u64 = 0;
        let mut depth: i32 = -1;
        while i < op_list.len() {
            match op_list[i] {
                "wtime" => {
                    if self.eng.board.color_to_move == WHITE {
                        self.timer.msec_left = op_list[i + 1].parse().unwrap();
                    }
                    i += 2;
                }
                "btime" => {
                    if self.eng.board.color_to_move == BLACK {
                        self.timer.msec_left = op_list[i + 1].parse().unwrap();
                    }
                    i += 2;
                }

                "winc" => {
                    if self.eng.board.color_to_move == WHITE {
                        self.timer.msec_inc = op_list[i + 1].parse().unwrap();
                    }
                    i += 2;
                }
                "binc" => {
                    if self.eng.board.color_to_move == BLACK {
                        self.timer.msec_inc = op_list[i + 1].parse().unwrap();
                    }
                    i += 2;
                }
                "depth" => {
                    depth = op_list[i + 1].parse().unwrap();
                    i += 2;
                }
                "nodes"
                => {
                    i += 2;
                }
                "infinite" => {
                    i += 1;
                }
                "movetime" => {
                    movetime = op_list[i + 1].parse().unwrap();
                    i += 2;
                }
                _ => {
                    i += 1;
                }
            }
        }

        self.timer.move_time = movetime;

        let max_time_think;
        if depth == -1 {
            depth = 20;
            max_time_think = self.timer.max_allocable().as_millis();
        } else {
            max_time_think = 0;
        }

        let res = self.eng.search(depth, max_time_think);

        let mov = res.1;
        let _score = res.0;


        println!("bestmove {}", mov.to_uci_string());
    }

    fn uci_handle_stop(&self) {
        panic!("NotImplementedError");
    }

    fn uci_handle_quit(&self) {
        std::process::exit(0);
    }
}

File: python_prototype/orchestradirector.py
import random

from engine import Engine
from move import Move
from board import Board
from timer import Timer
from typing import Optional, List
import helpers

DEBUG = False


# static class
class OrchestraDirector:
    board: Optional[Board] = None
    timer: Optional[Timer] = None

    @classmethod
    def init_startpos(cls):
        if DEBUG:
            print("startpos")
        cls.board = Board.from_startpos()

    @classmethod
    def handle_command(cls, command, options):
        match command:
            case "uci":
                cls.uci_handle_uci()
            case "isready":
                cls.uci_handle_isready()
            case "ucinewgame":
                pass
            case "position":
                cls.uci_handle_position(options)
            case "go":
                cls.uci_handle_go(options)
            case "stop":
                cls.uci_handle_stop()
            case "quit":
                cls.uci_handle_quit()
            case "setoption":
                pass
            case _:
                if DEBUG:
                    raise NotImplementedError(command, options)

    @classmethod
    def uci_handle_uci(cls):
        print("id name Orchestra")
        print("id author Dario & Mattia")
        print("uciok")

    @classmethod
    def uci_handle_position(cls, options):
        if options[0:8] == "startpos":
            cls.init_startpos()
            w = options.split("moves")
            if len(w) > 1:
                cls.execute_moves(w[1])
        else:
            fen, moves = helpers.split_fen_moves(options)  # not sure it goes there
            cls.init_from_fen(fen)
            cls.execute_moves(moves)

    @classmethod
    def execute_moves(cls, param):
        moves: List[str] = param.split()
        for mov_str in moves:
            mov: Move = Move.from_string(mov_str, cls.board)
            cls.board.make_move(mov)

    @classmethod
    def init_from_fen(cls, fen):
        # todo: review this because the string editing is done in two different places
        # fen = options[options.find("[") + 1:options.find("]")]
        cls.board = Board.from_fen(fen)

    @classmethod
    def uci_handle_isready(cls):
        print("readyok")

    @classmethod
    def uci_handle_go(cls, options):
        op_list = options.split()
        i = 0
        while i < len(op_list):
            match op_list[i]:
                case "wtime":
                    i += 2
                case "btime":
                    i += 2
                case "winc":
                    i += 2
                case "binc":
                    i += 2
                case "depth":
                    i += 2
                case "nodes":
                    i += 2
                case "infinite":
                    i += 1
                case _:
                    i += 1

        legal_moves = cls.board.generate_moves()

        eng = Engine(cls.board)
        mov = eng.search(6)[1]

        print("bestmove " + mov.to_string())
        return mov

    @classmethod
    def uci_handle_stop(cls):
        raise NotImplementedError

    @classmethod
    def uci_handle_quit(cls):
        exit(0)


if __name__ == '__main__':
    OrchestraDirector.handle_command("position",  "fen 8/8/8/K7/8/8/7Q/1k6 b - - 0 1 moves b1c1")
    # OrchestraDirector.handle_command("position", "startpos moves e2e4")
    OrchestraDirector.handle_command("go", "movetime 1000")
